### PR TITLE
COZMO-6600 Improve multi-device support.

### PR DESCRIPTION
**Pull request by gwatts**
Make the connectors multi-device aware by keeping track of which devices
the connector currently has an active connection to, and skipping those
when searching for another device.

Also add support for connecting to a device with an explicit serial
number, and log those serial numbers upon connection.

### CODE REVIEW

File: src/cozmo/run.py [hunk 1]
```diff
@@ -103,20 +113,50 @@ class IOSConnector(DeviceConnector):
...
+
+        else:
+            device_id = await self.usbmux.wait_for_serial(self.serial)
+            transport, proto = await self.usbmux.connect_to_device(
+                    protocol_factory, device_id, self.cozmo_port)
+
+        proto.device_info={
+            'device_type': 'ios',
+            'device_id': device_id,
+            'serial':  transport.device_info.get('SerialNumber'),
```
Comment (msintov) [inline src/cozmo/run.py]: The comma at the end of the 'serial' line could be removed

File: src/cozmo/usbmux/usbmux.py [hunk 2]
```diff
@@ -240,64 +248,119 @@ def msg_received(self, msg):
...
-        '''Open a TCP connection to the first device that has the requested port open.
+    async def wait_for_serial(self, serial, timeout=DEFAULT_MAX_WAIT):
+        '''Wait for a device with the specified serial number to attach.
 
         Args:
-            protocol_factory (callable): A callable that returns an asyncio.Protocol implementation
-            port (int): The port to connect to on the target device
-            max_wait (float): The maximum amount of time to wait for a suitable device to be connected
+            serial (string): Serial number of the device to wait for.
+            timeout (float): The maximum amount of time in secondsto wait for a
```
Comment (msintov) [inline src/cozmo/usbmux/usbmux.py]: Should be "in seconds to" (add a space)

File: src/cozmo/usbmux/usbmux.py [hunk 3]
```diff
@@ -240,64 +248,119 @@ def msg_received(self, msg):
...
+                matching device to be connected.
+                Set to None to wait indefinitely, or -1 to only check currently
+                connected devices.
         Returns:
-            (asyncio.Transport, asyncio.Protocol) - The connected transport and protocol
+            int: The device id of the connected device
         Raises:
-            asyncio.TimeoutError - If no devices with the requested port become available in the specified time
+            asyncio.TimeoutError - If the device with the specified serial number doesn't appear.
```
Comment (msintov) [inline src/cozmo/usbmux/usbmux.py]: Could be adjusted to remove the dash: "asyncio.TimeoutError if the device with the specified serial number doesn't appear."

File: src/cozmo/usbmux/usbmux.py [hunk 4]
```diff
@@ -240,64 +248,119 @@ def msg_received(self, msg):
...
+        '''Wait for the next device attachment event.
 
         Args:
             timeout (float): Maximum amount of time to wait for an event, or None for no timeout
         Returns:
-            int: The device id that was connected
+            int: The device id that attached.
         Raises:
-            asyncio.TimeoutError
+            asyncio.TimeoutError - If no devices with the requested port become
```
Comment (msintov) [inline src/cozmo/usbmux/usbmux.py]: Could be reworded to remove the dash: "asyncio.TimeoutError if no devices..."

File: src/cozmo/usbmux/usbmux.py [hunk 5]
```diff
@@ -324,3 +387,85 @@ def device_detached(self, device_id):
...
+        try:
+            return self._q.popleft()
+        except IndexError:
+            pass
+        self._wake = asyncio.Future(loop=self.loop)
+        await asyncio.wait_for(self._wake, loop=self.loop, timeout=timeout)
+        return self._q.popleft()
+
+
+class Timeout:
```
Comment (msintov) [inline src/cozmo/usbmux/usbmux.py]: Should there be a docstring for the Timeout class?

File: examples/multi_robot1.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/env python3
```
Comment (gwatts) [inline examples/multi_robot1.py]: renamed

File: src/cozmo/run.py (hunk 1 above)
Comment (gwatts) [inline src/cozmo/run.py]: fixed

File: src/cozmo/usbmux/usbmux.py (hunk 2 above)
Comment (gwatts) [inline src/cozmo/usbmux/usbmux.py]: fixed

File: src/cozmo/usbmux/usbmux.py (hunk 3 above)
Comment (gwatts) [inline src/cozmo/usbmux/usbmux.py]: fixed

File: src/cozmo/usbmux/usbmux.py (hunk 4 above)
Comment (gwatts) [inline src/cozmo/usbmux/usbmux.py]: fixed

File: src/cozmo/usbmux/usbmux.py (hunk 5 above)
Comment (gwatts) [inline src/cozmo/usbmux/usbmux.py]: added (though it's a small class intended to be internal to that module)

File: examples/multi_robot_independent.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    cozmo.logger.info("Turning robot 2")
+    await robot.turn_in_place(degrees(-90)).wait_for_completed()
+
+
+if __name__ == '__main__':
+    cozmo.setup_basic_logging()
+    loop = asyncio.get_event_loop()
+
+    # Connect to both robots
+    conn1 = cozmo.connect_on_loop(loop)
```
Comment (MarkWesleyAtAnki) [inline examples/multi_robot_independent.py]: Won't these examples spew an exception trace if there's nothing to connect to - maybe follow a similar pattern to that used in the other examples?

File: src/cozmo/run.py [hunk 8]
```diff
@@ -103,20 +113,50 @@ class IOSConnector(DeviceConnector):
...
+        serial (string): Serial number of the device to connect to.
+            If None, then connect to the first available iOS device running
+            the Cozmo app in SDK mode.
     '''
-    def __init__(self, **kw):
+    def __init__(self, serial=None, **kw):
         super().__init__(**kw)
         self.usbmux = None
+        self._connected = set()
+        self.serial = serial
```
Comment (MarkWesleyAtAnki) [inline src/cozmo/run.py]: "serial" doesn't necessarily imply a "serial number", could be e.g. misinterpreted as "is this serial (vs parallel)"

File: examples/multi_robot_independent.py (hunk 7 above)
Comment (gwatts) [inline examples/multi_robot_independent.py]: updated

File: src/cozmo/run.py (hunk 8 above)
Comment (gwatts) [inline src/cozmo/run.py]: I don't really see how it's going to be confused in the context its in, nor with the docstring that makes it explicit

File: src/cozmo/usbmux/usbmux.py [hunk 9]
```diff
@@ -248,23 +248,21 @@ def msg_received(self, msg):
             device_id (int): The id of the device to connect to
             port (int): The port to connect to on the target device
         Returns:
-            (asyncio.Transport, asyncio.Protocol): The connected transport and protocol.
-                An attribute called ``device_info`` will be set on the transport
-                holding a dictionary with more information about the device,
-                such as SerialNumber.
+            (dict, asyncio.Transport, asyncio.Protocol): The device infor,mation,
```
Comment (MarkWesleyAtAnki) [inline src/cozmo/usbmux/usbmux.py]: looks like a stray comma in infor,mation

**Pull request merged**